cai_causal_graph/identify_utils.py

```diff
@@ -90,46 +92,42 @@ def identify_confounders(graph: CausalGraph, node_1: NodeLike, node_2: NodeLike)
-        """Private function that does not check if DAG or do descendant pruning."""
+    def _identify_confounders_no_checks_no_descendant_pruning_networkx(
+        clean_graph: networkx.DiGraph, n1: str, n2: str
+    ) -> Set[str]:
+        """
+        Private function that does not check if DAG or do descendant pruning.
 
+        This function searches for confounders using the node `n1` as the starting point. Importantly, this
+        does not produce a set of minimal confounders.
+        """
         # create a copy of the provided graph and prune the children of node 1 and node 2
-        final_graph = clean_graph.copy()
-        for child in final_graph.get_children(n1):
-            final_graph.remove_edge(source=n1, destination=child)
-        for child in final_graph.get_children(n2):
-            final_graph.remove_edge(source=n2, destination=child)
+        removed_edges = list()
+        final_graph = clean_graph
+        for child in list(final_graph.successors(n1)):
+            removed_edges.append((n1, child))
+            final_graph.remove_edge(n1, child)
+        for child in list(final_graph.successors(n2)):
+            removed_edges.append((n2, child))
+            final_graph.remove_edge(n2, child)
 
         # search the parents of node 1 and check whether any of them is an ancestor to node 2
         confounders = set()
-        for parent in final_graph.get_parents(n1):
+        for parent in list(final_graph.predecessors(n1)):
             # add the parent to the confounding set if a directed path exists
-            if parent in final_graph.get_ancestors(n2):
+            if parent in networkx.ancestors(final_graph, n2):
                 confounders.add(parent)
             # otherwise, recursively call this function to identify confounders of the parent
             else:
                 confounders = confounders.union(
-                    set(_identify_confounders_no_checks_no_descendant_pruning(final_graph, parent, n2))
+                    set(_identify_confounders_no_checks_no_descendant_pruning_networkx(final_graph, parent, n2))
                 )
 
-        # do the reverse of the above by searching through the parents of node 2
-        confounders_reverse = set()
-        for parent in final_graph.get_parents(n2):
-            # add the parent to the confounding set if a directed path exists
-            if parent in final_graph.get_ancestors(n1):
-                confounders_reverse.add(parent)
-            # otherwise, recursively call this function to identify confounders of the parent
-            else:
-                confounders_reverse = confounders_reverse.union(
-                    set(_identify_confounders_no_checks_no_descendant_pruning(final_graph, parent, n1))
-                )
+        # add the edges that were removed back
+        for edge in removed_edges:
+            final_graph.add_edge(*edge)
```

Comment:
I see you add them back but we could just not do this if we copied. Or is this better as we use much less memory?

Reply:
Answered above :)